src/mesh_smoothing.rs

```diff
@@ -0,0 +1,582 @@
+use nalgebra::geometry::Point3;
+
+use crate::convert::cast_u32;
+use crate::geometry::Geometry;
+use crate::mesh_topology_analysis::vertex_to_vertex_topology;
+
+/// # Mesh relaxation / Laplacian smoothing
+/// Relaxes angles between mesh edges, resulting in a smoother geometry.
+/// The number of vertices, faces and the overall topology remains unchanged.
+/// The more iterations, the smoother result.
+/// Too many iterations may cause slow calculation time.
+///
+/// The algorithm is based on replacing each vertex position
+/// with an average position of its immediate neighbors
+#[allow(dead_code)]
+pub fn laplacian_smoothing(geometry: &Geometry, iterations: u8) -> Geometry {
+    laplacian_smoothing_with_anchors(geometry, iterations, &[])
+}
+
+/// # Mesh relaxation / Laplacian smoothing with fixed vertices
+/// Relaxes angles between mesh edges, while keeping some vertices anchored,
+/// resulting in a smoother geometry.
+/// The number of vertices, faces and the overall topology remains unchanged.
+/// The more iterations, the smoother result.
+/// Too many iterations may cause slow calculation time.
+///
+/// The algorithm is based on replacing each vertex position
+/// with an average position of its immediate neighbors
+#[allow(dead_code)]
+pub fn laplacian_smoothing_with_anchors(
+    geometry: &Geometry,
+    iterations: u8,
+    fixed_vertex_indices: &[u32],
+) -> Geometry {
+    let vertex_to_vertex_topology = vertex_to_vertex_topology(geometry);
+    let mut geometry_vertices: Vec<Point3<f32>> = Vec::from(geometry.vertices());
+    let mut vertices = geometry_vertices.clone();
+
+    for _ in 0..iterations {
+        for (current_vertex_index, neighbors_indices) in vertex_to_vertex_topology.iter() {
+            if fixed_vertex_indices
+                .iter()
+                .all(|i| *i != cast_u32(*current_vertex_index))
+            {
+                let mut average_position: Point3<f32> = Point3::origin();
+                for neighbor_index in neighbors_indices {
+                    average_position += geometry_vertices[*neighbor_index].coords;
+                }
+                average_position /= neighbors_indices.len() as f32;
+                vertices[*current_vertex_index] = average_position;
+            }
+        }
+        geometry_vertices = vertices.clone();
+    }
+
+    Geometry::from_triangle_faces_with_vertices_and_normals(
+        geometry.triangle_faces_iter().collect(),
+        vertices,
+        geometry.normals().to_vec(),
+    )
+}
+
+#[cfg(test)]
+mod tests {
+    use nalgebra::distance_squared;
+
+    use crate::edge_analysis::edge_valencies;
+    use crate::geometry::{Geometry, NormalStrategy, OrientedEdge, TriangleFace, Vertices};
+    use crate::mesh_analysis::border_vertex_indices;
+
+    use super::*;
+
```

Comment:
Can you please add a `FIXME: snapshot testing` here? We should at some point have a standard set of testing geometries for these things and not define them in code.